client/matching/loadbalancer.go

```diff
@@ -135,6 +140,144 @@ func (lb *defaultLoadBalancer) pickPartition(
 		return taskQueue.GetName()
 	}
 
-	n := max(1, nPartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
+	n := max(1, lb.nWritePartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
 	return tqName.WithPartition(rand.Intn(n)).FullName()
 }
+
+// PickReadPartition picks a partition for poller to poll task from, and keeps load balanced between partitions.
+// Caller is responsible to call pollToken.Release() after complete the poll.
+func (lb *defaultLoadBalancer) PickReadPartition(
+	namespaceID namespace.ID,
+	taskQueue taskqueuepb.TaskQueue,
+	taskQueueType enumspb.TaskQueueType,
+	forwardedFrom string,
+) *pollToken {
+	if forwardedFrom != "" || taskQueue.Kind == enumspb.TASK_QUEUE_KIND_STICKY {
+		// no partition for sticky task queue and forwarded request
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	parsedName, err := tqname.Parse(taskQueue.GetName())
+	if err != nil || err == nil && !parsedName.IsRoot() {
+		// parse error or partition already picked, use as-is
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	tqlb := lb.getTaskQueueLoadBalancer(namespaceID, parsedName, taskQueueType)
+
+	// For read path it's safer to return global default partition count instead of root partition, when we fail to
+	// map namespace ID to name.
+	var partitionCount = dynamicconfig.GlobalDefaultNumTaskQueuePartitions
```

Comment:
:(  I don't really like any more code being aware of this default (that we hope to change eventually). What's not safe about returning 0? Tasks would get forwarded there eventually?

Reply:
Tasks should forward at rate of 10 tasks per sec, but only if the root partition doesn't have a backlog.
This line should run very rarely (only when we fail to map NS name to ID). We can remove this later if needed with little/no trouble.